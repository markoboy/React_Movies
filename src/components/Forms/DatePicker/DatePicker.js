/* eslint-disable jsx-a11y/label-has-associated-control */
import withFormElementWrapper from '@components/HigherOrder/WithFormElementWrapper';
import { faCalendarAlt } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React, { memo } from 'react';
// eslint-disable-next-line import/no-named-default
import { default as ReactDatePicker } from 'react-date-picker/dist/entry.nostyle';
import FormLabel from '../FormLabel/FormLabel';
import './DatePicker.scss';

function DatePicker({
  value,
  onChange,
  required,
  label,
}) {
  return (
    <>
      <FormLabel label={label} />
      <ReactDatePicker
        value={value}
        onChange={onChange}
        calendarIcon={<FontAwesomeIcon icon={faCalendarAlt} />}
        clearIcon={null}
        format="dd-MM-y"
        dayPlaceholder="Day"
        monthPlaceholder="Month"
        yearPlaceholder="Year"
        required={required}
      />
    </>
  );
}

DatePicker.defaultProps = {
  onChange: null,
  required: false,
  label: '',
};

DatePicker.propTypes = {
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  label: PropTypes.string,
};

export default memo(withFormElementWrapper(DatePicker));
