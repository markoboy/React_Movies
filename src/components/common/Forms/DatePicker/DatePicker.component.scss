@import 'essential', 'react-date-picker/dist/DatePicker.css', 'react-calendar/dist/Calendar.css';

.react-date-picker {
  width: 100%;
  color: $color-primary;
  cursor: pointer;
  border: 1px solid transparent;
  border-radius: $border-radius;

  &:focus,
  &:hover {
    border-color: $color-primary;
    box-shadow: $box-shadow-primary;
  }

  input {
    color: $color-primary;
    background-color: transparent;
    font-size: $form-input-font-size;

    &:invalid {
      background-color: transparent;
    }

    &::placeholder {
      color: $color-secondary;
      opacity: 1;
    }
  }

  &__calendar-button {
    color: $color-primary;
    padding: $form-input-padding;
  }

  &__wrapper {
    border: none;
    background-color: rgba($color-secondary, 0.5);
    border-radius: $border-radius;
  }

  &__inputGroup {
    padding: $form-input-padding;
  }
}
