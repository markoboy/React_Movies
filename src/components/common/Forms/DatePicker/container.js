import React, { useCallback, useRef } from 'react';
import DatePickerComponent from './component';

const flatpickrOptions = {
  dateFormat: 'd-M-Y',
  mode: 'single',
};

const DatePicker = ({ onChange, ...props }) => {
  const flatpickrInstance = useRef();

  const handleOnCreate = useCallback((flatpickr) => {
    flatpickrInstance.current = flatpickr;
  }, []);

  const handleOnDestroy = useCallback(() => {
    flatpickrInstance.current = null;
  }, []);

  const formatDate = useCallback((defaultDateValue, dateValue) => {
    if (typeof dateValue === 'string') {
      return dateValue;
    }

    const formatedDate = flatpickrInstance.current?.formatDate(
      dateValue,
      flatpickrOptions.dateFormat
    );

    return formatedDate || defaultDateValue || '';
  }, []);

  const handleDateChange = useCallback(
    (dates) => {
      onChange?.(dates[0]);
    },
    [onChange]
  );

  return (
    <DatePickerComponent
      onChange={handleDateChange}
      onCreate={handleOnCreate}
      onDestroy={handleOnDestroy}
      formatDate={formatDate}
      options={flatpickrOptions}
      {...props}
    />
  );
};

export default DatePicker;
