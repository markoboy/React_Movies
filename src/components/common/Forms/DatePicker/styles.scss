@import 'essential', 'flatpickr/dist/themes/dark.css';

.flatpickr-calendar {
  box-shadow: $box-shadow;
}

// Change the background color of the calendar container
.flatpickr-calendar,
.flatpickr-months .flatpickr-month,
.flatpickr-current-month .flatpickr-monthDropdown-months,
span.flatpickr-weekday {
  background-color: $color-grey;
}

.flatpickr-day.selected, .flatpickr-day.startRange, .flatpickr-day.endRange, .flatpickr-day.selected.inRange, .flatpickr-day.startRange.inRange, .flatpickr-day.endRange.inRange, .flatpickr-day.selected:focus, .flatpickr-day.startRange:focus, .flatpickr-day.endRange:focus, .flatpickr-day.selected:hover, .flatpickr-day.startRange:hover, .flatpickr-day.endRange:hover, .flatpickr-day.selected.prevMonthDay, .flatpickr-day.startRange.prevMonthDay, .flatpickr-day.endRange.prevMonthDay, .flatpickr-day.selected.nextMonthDay, .flatpickr-day.startRange.nextMonthDay, .flatpickr-day.endRange.nextMonthDay {
  background-color: $color-primary;
  border-color: $color-primary;
}

// Change the hover/focus effect of the calendar
.flatpickr-day.inRange, .flatpickr-day.prevMonthDay.inRange, .flatpickr-day.nextMonthDay.inRange, .flatpickr-day.today.inRange, .flatpickr-day.prevMonthDay.today.inRange, .flatpickr-day.nextMonthDay.today.inRange, .flatpickr-day:hover, .flatpickr-day.prevMonthDay:hover, .flatpickr-day.nextMonthDay:hover, .flatpickr-day:focus, .flatpickr-day.prevMonthDay:focus, .flatpickr-day.nextMonthDay:focus {
  background-color: $color-black;
  border-color: $color-black;
}

// Change the chevron right/left fill color
.flatpickr-months .flatpickr-prev-month:hover svg, .flatpickr-months .flatpickr-next-month:hover svg {
  fill: $color-primary;
}

.date-picker {
  &__calendar-button {
    position: absolute;
    bottom: 20px;
    right: 15px;
    color: $color-primary;
    pointer-events: none;
  }

}
