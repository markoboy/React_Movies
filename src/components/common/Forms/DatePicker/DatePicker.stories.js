import React from 'react';

import withStoryBackgroundWrapper from '@components/hocs/WithStoryBackgroundWrapper';
import DatePicker from './index';

export default {
  component: DatePicker,
  title: 'Common/Forms/DatePicker',
  decorators: [withStoryBackgroundWrapper],
};

const Template = (args) => <DatePicker {...args} />;

export const Default = Template.bind({});
Default.args = {
  placeholder: 'Choose a date',
};

export const WithLabel = Template.bind({});
WithLabel.args = {
  label: 'Date',
  placeholder: 'Choose a date',
};
