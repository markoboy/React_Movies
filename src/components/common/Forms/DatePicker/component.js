import Input from '@components/common/Forms/Input';
import { faCalendarAlt } from '@fortawesome/free-solid-svg-icons/faCalendarAlt';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import React from 'react';
import Flatpickr from 'react-flatpickr';
import './styles.scss';

const noop = () => {};

function DatePickerComponent({
  id,
  name,
  defaultValue,
  required,
  label,
  placeholder,
  options,
  onChange,
  onCreate,
  onDestroy,
  formatDate,
}) {
  return (
    <Flatpickr
      value={defaultValue}
      onChange={onChange}
      onCreate={onCreate}
      onDestroy={onDestroy}
      options={options}
      // Render props
      render={({ value: dateValue, defaultValue: defaultDate }, ref) => (
        <Input
          inputRef={ref}
          id={id}
          name={name}
          value={formatDate(defaultDate, dateValue)}
          label={label}
          placeholder={placeholder}
          required={required}
          // Required else an error about read-only field
          onChange={noop}
        >
          <FontAwesomeIcon
            className="date-picker__calendar-button"
            icon={faCalendarAlt}
          />
        </Input>
      )}
    />
  );
}

DatePickerComponent.defaultProps = {
  required: false,
  label: '',
  defaultValue: '',
};

DatePickerComponent.propTypes = {
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  required: PropTypes.bool,
  label: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};

export default DatePickerComponent;
