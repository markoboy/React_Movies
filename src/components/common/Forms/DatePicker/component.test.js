import React from 'react';
import { render } from '@testing-library/react';

import DatePicker from './component';

jest.mock('react-date-picker/dist/entry.nostyle', () => () => 'Mock Component');
const mockFn = jest.fn();

describe('DatePicker Component', () => {
  it('renders DatePicker Component with provided label', () => {
    const { container, getByText } = render(
      <DatePicker onChange={mockFn} label="Date picker" />
    );

    const datePicker = getByText(/Date picker/);

    expect(datePicker).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });
});
