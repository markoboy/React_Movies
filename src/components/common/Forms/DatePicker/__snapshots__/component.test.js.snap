// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker Component renders DatePicker Component with provided label 1`] = `<div />`;
